prometheus_keys.lua

```diff
@@ -0,0 +1,128 @@
+-- Storage to keep track of used keys. Allows to atomically create, delete
+-- and list keys. The keys are synchronized between nginx workers
+-- using ngx.shared.dict. The whole purpose of this module is to avoid
+-- using ngx.shared.dict:get_keys (see https://github.com/openresty/lua-nginx-module#ngxshareddictget_keys),
+-- which blocks all workers and therefore it shouldn't be used with large
+-- amounts of keys.
+
+local KeyIndex = {}
+KeyIndex.__index = KeyIndex
+
+local lock_lib = require("prometheus_lock")
+
+function KeyIndex.new(shared_dict, prefix)
+  local self = setmetatable({}, KeyIndex)
+  self.dict = shared_dict
+  self.key_prefix = prefix .. "key_"
+  self.sync_hint_prefix = prefix .. "sync_"
+  self.key_count = prefix .. "key_count"
+  self.last = 0
+  self.keys = {}
+  self.index = {}
+  self.lock = lock_lib.new(prefix .. "lock_keys", self.dict)
+  return self
+end
+
+-- Loads new keys that might have been added by other workers since last sync.
+function KeyIndex:sync()
+  local full_sync_hint = self.dict:get(self.sync_hint_prefix .. ngx.worker.id()) or 0
+  local N = self.dict:get(self.key_count) or 0
+  if full_sync_hint > 0 then
+    -- Some other worker deleted something, lets do a full sync.
+    self:sync_range(0, N)
+  elseif N ~= self.last then
+    -- Sync only new keys, if there are any.
+    self:sync_range(self.last, N)
+  end
+  return N
+end
+
+-- Iterates keys from first to last, adds new items and removes deleted items.
+function KeyIndex:sync_range(first, last)
+  for i = first, last do
+    -- Read i-th key. If it is nil, it means it was deleted by some other thread.
+    local key = self.dict:get(self.key_prefix .. i)
+    if key then
+      self.keys[i] = key
+      self.index[key] = i
+    elseif self.keys[i] then
+      self.index[self.keys[i]] = nil
+      self.keys[i] = nil
+    end
+  end
+end
+
+-- Sets timer to sync the index every interval seconds.
```

Comment:
Why is this necessary given that we trigger a sync anyway in `add` and `list`?

Reply:
The idea was to put an upper bound to the time the local key_index might be out of sync. But you're probably right that it doesn't matter, since it would only be important when we `collect` metrics and there we call `sync` anyway.

Reply:
Removed in 570964b4adf41b42dc10470698d84ca293412992.